File: services/api/alertType.ts
import BaseApi from "./base";
import { UUID } from "../../models/base";
import { AlertTypeCreate } from "../../models/alertType";

const api = new BaseApi();

export default {
  endpoint: "/alert/type",

  // CREATE
  async createAlertType(alertType: AlertTypeCreate) {
    return await api
      .createRequest(`${this.endpoint}`, alertType)
      .catch((err) => {
        throw err;
      });
  },

  // READ
  async getAllAlertTypes() {
    return await api.readRequest(`${this.endpoint}`).catch((err) => {
      throw err;
    });
  },

  async getAlertType(uuid: UUID) {
    return await api.readRequest(`${this.endpoint}${uuid}`).catch((err) => {
      throw err;
    });
  },

  // UPDATE
  async updateAlertType(alertType: { value: string }, uuid: UUID) {
    return await api
      .updateRequest(`${this.endpoint}${uuid}`, alertType)
      .catch((err) => {
        throw err;
      });
  },
};

File: backend/app/api/routes/alert_type.py
from fastapi import APIRouter, Depends, Request, Response
from fastapi_pagination.ext.sqlalchemy_future import paginate
from sqlalchemy.orm import Session
from sqlalchemy.sql.expression import select
from uuid import UUID

from api.models.alert_type import AlertTypeCreate, AlertTypeRead, AlertTypeUpdate
from api.routes import helpers
from db import crud
from db.database import get_db
from db.schemas.alert_type import AlertType


router = APIRouter(
    prefix="/alert/type",
    tags=["Alert Type"],
)


#
# CREATE
#


def create_alert_type(
    create: AlertTypeCreate,
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    obj: AlertType = crud.create(obj=create, db_table=AlertType, db=db)

    response.headers["Content-Location"] = request.url_for("get_alert_type", uuid=obj.uuid)


helpers.api_route_create(router, create_alert_type)


#
# READ
#


def get_all_alert_types(db: Session = Depends(get_db)):
    return paginate(db, select(AlertType).order_by(AlertType.value))


def get_alert_type(uuid: UUID, db: Session = Depends(get_db)):
    return crud.read(uuid=uuid, db_table=AlertType, db=db)


helpers.api_route_read_all(router, get_all_alert_types, AlertTypeRead)
helpers.api_route_read(router, get_alert_type, AlertTypeRead)


#
# UPDATE
#


def update_alert_type(
    uuid: UUID,
    alert_type: AlertTypeUpdate,
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    crud.update(uuid=uuid, obj=alert_type, db_table=AlertType, db=db)

    response.headers["Content-Location"] = request.url_for("get_alert_type", uuid=uuid)


helpers.api_route_update(router, update_alert_type)


#
# DELETE
#


def delete_alert_type(uuid: UUID, db: Session = Depends(get_db)):
    crud.delete(uuid=uuid, db_table=AlertType, db=db)


helpers.api_route_delete(router, delete_alert_type)
